aries_vcx/docs/guidelines.md

```diff
@@ -0,0 +1,71 @@
+# Guidelines for implementing protocol state machines
+
+## No 'Sent' states
+Do not create states to track whether a reply has been sent to counterparty. It is responsibility of
+`aries-vcx` consumers to keep track of whether response message has been sent to counterparty. From state
+machine perspective, creation of a protocol message implies a reply to that message can be processed.
+
+Example: Do not create states such as `PresentationSent` (in context of `presentation-protocol` in role of prover).
+Instead, only create state `PresentationPrepared` which enables `aries-vcx` consumers to get the presentation message
+and send it to the counterparty. Subsequently, it should be possible to further drive state machine in  `PresentationPrepared` 
+using `presentation-ack` message received from verifier. 
+
+## Separate final states
+While Aries RFCs commonly refer to single final state to convey nothing more can happen after exchange
+of certain messages - for example in diagram [here](https://github.com/hyperledger/aries-rfcs/blob/main/features/0453-issue-credential-v2/README.md)
+- on implementation level we want to distinguish different internal states signalling different circumstances (which 
+simply happen to share the property of being final).
+
+Example: instead of having `Final` state which would maintain `status: Success | Failed | Rejected`, we
+should have 3 separate states `Success`, `Failed`, `Rejected`.
+
+## Receives Aries messages always move state
+Processing a message from counterparty must move state. Even if we are dealing just with acknowledgement type of message.
```

Comment:
IMO the claim in the heading is too strong (confusingly so):
* just receiving a message by itself shouldn't move state, but processing it should,
* this obviously doesn't apply to final states (so "always" is too much).
Perhaps an alternative (although still imperfect) would be something like: "Protocol-based expectation of message receipt implies non-final state".

Reply:
While I addressed your comment, I've reached for simpler wording - let me know if you think thins works